pallets/liquidity-mining/src/lib.rs

```diff
@@ -0,0 +1,1325 @@
+				if !g_pool.total_shares_z.is_zero() && g_pool.updated_at != now_period {
+					let reward_per_period = Self::get_global_pool_reward_per_period(
+						g_pool.yield_per_period.into(),
+						g_pool.total_shares_z,
+						g_pool.max_reward_per_period,
+					)?;
+					Self::update_global_pool(g_pool, now_period, reward_per_period)?;
+				}
+				g_pool.liq_pools_count = g_pool.liq_pools_count.checked_add(1).ok_or(Error::<T>::Overflow)?;
+
+				let liq_pool_id = Self::get_next_id()?;
+				let pallet_account = Self::account_id();
+				let (nft_class, _) = pallet_nft::Pallet::<T>::do_create_class(
+					pallet_account,
+					ClassType::LiquidityMining,
+					vec![].try_into().unwrap(),
+				)?;
+				<NftClassData<T>>::insert(nft_class, (asset_pair, 0, g_pool.id));
+
+				let pool = LiquidityPoolYieldFarm::new(liq_pool_id, now_period, loyalty_curve, multiplier, nft_class);
+
+				<LiquidityPoolData<T>>::insert(g_pool.id, &amm_pool_id, &pool);
+
+				Self::deposit_event(Event::LiquidityPoolAdded(g_pool.id, amm_pool_id, pool));
+
+				Ok(())
+			})
+		}
+
+		#[pallet::weight(1000)]
+		#[transactional]
+		pub fn update_liquidity_pool(
+			origin: OriginFor<T>,
+			farm_id: PoolId,
+			asset_pair: AssetPair,
+			multiplier: PoolMultiplier,
+		) -> DispatchResult {
+			//TODO: this fn is WIP, it's not tested
+			let who = ensure_signed(origin)?;
+
+			ensure!(!multiplier.is_zero(), Error::<T>::InvalidMultiplier);
+
+			let amm_pool_id = T::AMM::get_pair_id(asset_pair);
+			<LiquidityPoolData<T>>::try_mutate(farm_id, &amm_pool_id, |liq_pool| {
+				let liq_pool = liq_pool.as_mut().ok_or(Error::<T>::LiquidityPoolNotFound)?;
+
+				<GlobalPoolData<T>>::try_mutate(farm_id, |g_pool| {
+					let g_pool = g_pool.as_mut().ok_or(Error::<T>::FarmNotFound)?;
+
+					ensure!(who == g_pool.owner, Error::<T>::Forbidden);
+
+					let now_period = Self::get_now_period(g_pool.blocks_per_period)?;
+
+					let reward_per_period = Self::get_global_pool_reward_per_period(
+						g_pool.yield_per_period.into(),
+						g_pool.total_shares_z,
+						g_pool.max_reward_per_period,
+					)?;
+					Self::update_global_pool(g_pool, now_period, reward_per_period)?;
+
+					let pool_reward = Self::claim_from_global_pool(g_pool, liq_pool, liq_pool.stake_in_global_pool)?;
+					Self::update_liq_pool(liq_pool, pool_reward, now_period, g_pool.id, g_pool.reward_currency)?;
+
+					let incentivized_token_balance_in_amm =
+						T::MultiCurrency::free_balance(g_pool.reward_currency, &amm_pool_id);
+
+					let new_stake_in_global_pool = incentivized_token_balance_in_amm
+						.checked_mul(liq_pool.total_valued_shares)
+						.ok_or(Error::<T>::Overflow)?
+						.checked_mul(multiplier.into())
+						.ok_or(Error::<T>::Overflow)?;
+
+					g_pool
+						.total_shares_z
+						.checked_sub(liq_pool.stake_in_global_pool)
+						.ok_or(Error::<T>::Overflow)?;
+					g_pool
+						.total_shares_z
+						.checked_add(new_stake_in_global_pool)
+						.ok_or(Error::<T>::Overflow)?;
```

Comment:
i am bit confused of these calculations. What purpose they have since the result is not used ?